### PR TITLE
MNTOR-24: Feature Flags ADR

### PR DESCRIPTION
**Pull request by mansaj**
# References: 
Jira: MNTOR-24

# Description
We need a robust way to A/B testing features, manage when/what to deploy, and standardize feature flag practices to potentially service multiple code bases / projects in the future

### CODE REVIEW

File: docs/adr/0005-feature-flags.md
Comment (pdehaan) [inline docs/adr/0005-feature-flags.md:54]: Not sure if these align with the example above.
Should it be this?
1. `section`: &hellip;
2. `featurepurpose`: &hellip;
3. `service`: &hellip;

Or, should `section_featurepurpose_service` change? (if we do change it, "featurepurpose" feels a bit of a mouthful).

Comment (mansaj) [reply docs/adr/0005-feature-flags.md]: Thanks Peter! Changed

File: docs/adr/0005-feature-flags.md
Comment (Vinnl) [inline docs/adr/0005-feature-flags.md:52]: Could you clarify whether this is an exhaustive list? I'm always struggling figuring out what a "section" is, e.g. is it the settings page? Or a specific section of the settings page, such as Premium settings? What if a feature is available to both premium and free subscribers?

Alternatively, I'd also be fine with just skipping `section`, to avoid [smurf naming](https://blog.codinghorror.com/new-programming-jargon/#21).

File: docs/adr/0005-feature-flags.md
Comment (Vinnl) [inline docs/adr/0005-feature-flags.md:64]: Yes! Thank you for including this :clap: One other radical suggestion I've seen is to disallow feature flags enabling something for everyone, i.e. you can roll out to 95% of users, but to roll out to everyone, you need to remove the flag — but that's probably a bit too radical. :sweat_smile:

One thing that's been helpful on Relay is to define a union of strings [enumerating all active feature flags](https://github.com/mozilla/fx-private-relay/blob/57dde8ff1c5794330232e5ee899870e79facbab1/frontend/src/hooks/api/runtimeData.ts#L6-L16) in TypeScript. Then we can just remove a flag from that type, and TS will tell us which bits of code will need to be updated - because it's often not entire if/else branches you can just delete.

File: docs/adr/0005-feature-flags.md
Comment (Vinnl) [inline docs/adr/0005-feature-flags.md:86]:
```suggestion
* [Unleash](https://www.getunleash.io/) Open Source Feature Flag (self host)
* Unleash (managed cloud)
```

**Review by Vinnl — APPROVED** [review]
Looking good!

File: docs/adr/0005-feature-flags.md
Comment (flozia) [inline docs/adr/0005-feature-flags.md:78]: Having the ability to tag would be really nice.

File: docs/adr/0005-feature-flags.md
Comment (flozia) [inline docs/adr/0005-feature-flags.md:83]: Another option could be having a JSON file with a data structure that we define. When it comes to infrastructure that option would be similar to an `.env`, but it could give us still the ability to add more context like tagging, sub-feature flags and  dependencies on other features. Something like:
```
{
  "flags": [
    {
      "name": "Party mode",
      "key": "party-mode",
      "party-mode": {
        "dev": true,
        "prod": false
      },
      "dependencies": ["confetti"]
    },
    {
      "name": "Confetti",
      "key": "confetti",
      "party-mode": {
        "dev": true,
        "prod": false
      },
      "dependencies": []
    }
  ]
}
```

**Review by flozia — APPROVED** [review]

File: docs/adr/0005-feature-flags.md
Comment (maxxcrawford) [inline docs/adr/0005-feature-flags.md:114]: thought: This may also be a good security point, but limiting access to who can flip flags.

File: docs/adr/0005-feature-flags.md
Comment (maxxcrawford) [inline docs/adr/0005-feature-flags.md:117]: Question(s): Are other projects using this service? Has this gone through a vendor security review?

File: docs/adr/0005-feature-flags.md
Comment (maxxcrawford) [inline docs/adr/0005-feature-flags.md:128]: Question: What would a "Unleash" outage look like? Do all the fences go down and the dinosaurs escape?

File: docs/adr/0005-feature-flags.md
Comment (maxxcrawford) [inline docs/adr/0005-feature-flags.md:142]: Question: What does upkeep look like? Making sure to install security patches/updates?

File: docs/adr/0005-feature-flags.md
Comment (maxxcrawford) [inline docs/adr/0005-feature-flags.md:149]: Question: Would this power Opt 1 / Environment Variable?

**Review by maxxcrawford — COMMENTED** [review]
Mostly questions from my end. Great work, @mansaj!

**Review by rhelmer — APPROVED** [review]

File: docs/adr/0005-feature-flags.md
Comment (mansaj) [reply docs/adr/0005-feature-flags.md:52]: When naming something, I always think about future proofing, and namespaces. To namespace, I think we always want to think from large concepts to smaller concepts.

I don't know if "section" is the best way to word to describe what I was thinking there.. but I was thinking more about the different "versions" an app can exist in: free version, premium version, admin version (potentially)

File: docs/adr/0005-feature-flags.md
Comment (mansaj) [reply docs/adr/0005-feature-flags.md:83]: I agree, this is also a good DIY option. A lot more extensible. I'll make a separate option for this

Comment (mansaj) [reply docs/adr/0005-feature-flags.md:88]: Added `JSON` as an option. Thanks @flozia

File: docs/adr/0005-feature-flags.md
Comment (mansaj) [reply docs/adr/0005-feature-flags.md:120]: Agreed

File: docs/adr/0005-feature-flags.md
Comment (Vinnl) [inline docs/adr/0005-feature-flags.md:124]: What's unclear to me from this description is where this JSON would live? In an env var, or in a file in the repo? If the latter, how easy would it be to have different flags per environment? And then I'd also list as a downside that flipping a flag (e.g. if it introduces a bug that we quickly want to back out) would require a new hotfix release and deploy, which is probably riskier than just redeploying the same code with a different environment variable.

Comment (Vinnl) [reply docs/adr/0005-feature-flags.md:124]: Oh, I suppose the `environments` property adapts it to different environments. Shooting from the hip: maybe we could even add an `envVar` property that looks for a particular value in a feature-flag-specific env var, and then enable (or disable?) it when that value is present? Then we'd have the best of both worlds.

File: docs/adr/0005-feature-flags.md
Comment (mansaj) [reply docs/adr/0005-feature-flags.md:152]: Not within Mozilla, but they have a list on their website. 
After tech meeting, closing this discussion as we will no longer consider this option

File: docs/adr/0005-feature-flags.md
Comment (mansaj) [reply docs/adr/0005-feature-flags.md:163]: Yes, all dinosaurs would escape. I'd say part of our website will likely be down also. Not great

File: docs/adr/0005-feature-flags.md
Comment (mansaj) [reply docs/adr/0005-feature-flags.md:177]: Probably keeping an eye on it from time to time :( 
Pay more attention for major version upgrades

File: docs/adr/0005-feature-flags.md
Comment (mansaj) [reply docs/adr/0005-feature-flags.md:184]: No longer considering iron-session, but it was a good article

File: docs/adr/0005-feature-flags.md
Comment (mansaj) [reply docs/adr/0005-feature-flags.md:124]: Hmm good question. I think if it's JSON, the assumption is that the JSON would live in the code.

We definitely need to think about the pros and cons of these two approaches a bit more...
cc @rhelmer @maxxcrawford @flozia @codemist 

1. `env var` approach seems pretty straightforward, but it's actually not in practice:
    1. we actually have a hard time as devs (even SREs) to have a holistic picture of what are the env vars that are currently in place for stage and production environment. These are values that are buried deep inside of config maps and yaml files in moz services repo, which I don't know how many of our team has access
    2. changing `stage` and `prod` env vars require SRE approval / restart the server
2. let's think about the deployment of `json` approach
    1. changes to `stage` can be done by devs via our normal deployment process
    2. changes to `prod` requires SRE approval / restart
    3. I agree with you that the changes would be riskier since it'll be code changes -- but I guess if we do feature flags correctly, feature flags are meant for reducing this exact risk?
    4. we (at least as devs) will have a better view of what's in dev/stage/prod, re the first point under `env var`
    5. Not to mention all the other config benefits from having a json structure already in the doc
    

Separately, I'm thinking maybe the below format makes more sense, a bit more dup in data, but faster retrieval..:
```json
{
  "dev": {
    "name": "Premium Subplat Email", 
    "description": "description",
    "key": "premium_subplat_email",
    "createdAt": "",
    "expiresAt": "",
    ...
  }
  "stage":
  "prod":
}
```

Comment (Vinnl) [reply docs/adr/0005-feature-flags.md:124]: Regarding

> we actually have a hard time as devs (even SREs) to have a holistic picture of what are the env vars that are currently in place for stage and production environment. 

on Relay, we just expose the feature flag's env var's value for inspection ([at `/runtime_data/`](https://relay.firefox.com/api/v1/runtime_data)).

As for

> I agree with you that the changes would be riskier since it'll be code changes -- but I guess if we do feature flags correctly, feature flags are meant for reducing this exact risk?

I guess the question is if we're OK with an always-deployable `main` (in which case we could just push the latest `main` through to production with the feature flag flipped), or if we still want to make sure that commits since the latest deployments are not going to prod. If the former, the risk is indeed not increased, but if the latter, I'd at least mention it as a downside, though I wouldn't necessarily consider it reason not to go ahead with it.

Agreed with the rest of your analysis and your proposal.

Comment (mansaj) [reply docs/adr/0005-feature-flags.md:124]: @Vinnl Good points. 

Some of these feature flag discussions inevitably have to turn into "deployment process" discussions. For Monitor, so far, we haven't gone beyond the "always deployable main" model yet. We did have some tech meetings about potentially adopting a release branch strategy.. which escalates to having a dedicated "release engineer" or someone who is going to be in charge of cherry picking at the time of release.. which escalates to longer release cycles. I think for the size that we are at, the "always deployable main" strategy still seems to work, though I do have to acknowledge that the more people we have, the more chaotic `main` becomes

Comment (Vinnl) [reply docs/adr/0005-feature-flags.md:124]: Oh heh, I'd much prefer the always-deployable `main`, I just wasn't sure if we were there yet :sweat_smile:

File: docs/adr/0005-feature-flags.md
Comment (pdehaan) [inline docs/adr/0005-feature-flags.md:116]: Q: Does this allow us to do %-based rollouts?
Recently in FxA we did a 15% rollout of some new React work before rolling out to 100% of users.

**Pull request merged**
